src/cas/spreadsheet_to_cas.py

```diff
@@ -105,43 +155,73 @@ def spreadsheet2cas(
+    dataset_anndata, matrix_file_id = load_or_fetch_anndata(
+        anndata_file_path, meta_data_result
     )
-    if not anndata_file_path:
-        download_dataset_with_id(matrix_file_id)
-    dataset_anndata = read_anndata_file(anndata_file_path)
 
-    labelsets = OrderedDict()
+    cas = initialize_cas_structure(matrix_file_id, meta_data_result)
 
-    # metadata
-    cas = {
-        "matrix_file_id": matrix_file_id,
-        "cellannotation_schema_version": "TBA",
-        "cellannotation_timestamp": "TBA",
-        "cellannotation_version": "TBA",
-        "cellannotation_url": meta_data_result["CxG LINK"],
-        "author_name": "TBA",
-        "author_contact": "TBA",
-        "orcid": "TBA",
-        "annotations": [],
-        "labelsets": [],
-    }
+    labelsets = process_and_add_annotations(cas, dataset_anndata, raw_data_result)
+
+    process_and_add_labelsets(cas, labelset_list, labelsets)
 
-    # annotations
+    # Write the JSON data to the file
+    with open(output_file_path, "w") as json_file:
+        json.dump(cas, json_file, indent=2)
+
+
+def process_and_add_labelsets(
+    cas: dict, labelset_list: Optional[List[str]], labelsets: OrderedDict
+):
+    """
+    Adds labelsets with ranks to the CAS structure, using either a provided list or keys from `labelsets`.
+
+    Args:
+        cas (dict): The CAS structure to update.
+        labelset_list (Optional[List[str]]): Custom list to define labelset ranks. If None, uses `labelsets` keys.
+        labelsets (OrderedDict): Processed labelsets, used if no custom list is provided.
+
+    Uses `calculate_labelset_rank` to determine ranks, appending each labelset to CAS with its rank.
+    """
+    labelset_rank_dict = calculate_labelset_rank(
+        labelset_list if labelset_list else list(labelsets.keys())
+    )
+    for labelset, rank in labelset_rank_dict.items():
+        cas.get("labelsets").append(
+            {"name": labelset, "description": None, "rank": str(rank)}
+        )
+
+
+def process_and_add_annotations(cas, dataset_anndata, raw_data_result):
+    """
+    Adds processed annotations from raw data to the CAS structure and tracks labelsets.
+    Assumes certain external definitions for column names and transformation functions.
+
+    Args:
+        cas (dict): The CAS structure to update with annotations.
+        dataset_anndata (AnnData): Dataset for annotation context.
+        raw_data_result (DataFrame): Raw annotation data.
+
+    Returns:
+        OrderedDict: Tracks labelsets encountered, initialized to None.
+
+    Note:
+        Requires `custom_lowercase_transform`, `get_cell_ids`, and column constants to be defined.
+    """
     stripped_data_result = raw_data_result.map(
         lambda x: x.strip() if isinstance(x, str) else x
     )
+    labelsets = OrderedDict()
     for index, row in stripped_data_result.iterrows():
         labelset = row[LABELSET_COLUMN]
```

Comment:
According to spec "columns names MUST match CAS annotation object key names". We don't need these mappings in the code, they should match by name. 

Before the loop we can read the headers in the spreadsheet and match with the fields in the CAS schema. Then we can then use this matching while extracting data from the rows. We can update the test data to match column names with the schema fields.